### ISSUE TITLE
[Truffle] Super passing wrong block

### ISSUE THREAD

**Issue by nirvdrum** [jruby/jruby#2914]
The tests for [hike](https://rubygems.org/gems/hike) have a failure because we're not passing the correct block to `super`.  I've extracted a smaller snippet to demonstrate the problem:

```
class UppercaseArray < Array
  def collect!
    super do |element|
      result = yield element
      normalize_element(result)
    end
  end

  def normalize_element(element)
    element.upcase
  end
end

a = UppercaseArray.new
a.push "a", "b", "c"
a.collect! { |x| x.downcase }

p a
```

This should print `["A", "B", "C"]`, but instead prints `["a", "b", "c"]`.

The block from the `collect!` call (i.e., the one with `x.downcase`) is being passed to `super` and the block attached to the `super` call is completely ignored.

**Comment by nirvdrum** [comment]
If there's not a RubySpec covering the behavior, we should look at adding one.

**Comment by nirvdrum** [comment]
I think this is a variation of the same problem (this time take from [mime-types](https://rubygems.org/gems/mime-types)):

```
class HashWithArrayDefault < Hash
  def initialize
    super { |h, k| h[k] = [] }
  end
end

h = HashWithArrayDefault.new

p h[:not_exists] == []
```

This should print `true`, but instead prints `false`.

**Comment by chrisseaton** [comment]
Fixed in 36fbd3b576383d698efdd119d7978dd8a2e725c8

**Comment by chrisseaton** [comment]
Spec in d15ad1d3cf2572b8092e33003239e08a9755d13b.
